pyfar/signals/deterministic.py

```diff
@@ -256,6 +393,491 @@ def exponential_sweep_time(n_samples, frequency_range, n_fade_out=90,
+    # group delay at 0 Hz must be 0
+    sweep_gd = np.zeros(n_bins)
+    # group delay at df equals starting time unless it's 0
+    if start_margin > 0:
+        sweep_gd[1] = start_margin / sampling_rate
+        tg_start = 2
+    else:
+        tg_start = 1
+    # group delay at Nyquist equals stopping time
+    sweep_gd[-1] = (n_samples - stop_margin) / sampling_rate
+
+    # FORMULA (11, p.40, in Müller and Massarani 2001)
+    sweep_power = np.sum(np.abs(sweep_abs**2))
+    C = (sweep_gd[-1] - sweep_gd[1]) / sweep_power
+
+    # FORMULA (10, p.40, in Müller and Massarani 2001)
+    for k in range(tg_start, n_bins):  # index 2 to nyq
+        sweep_gd[k] = sweep_gd[k-1] + C * np.abs(sweep_abs[k])**2
+
+    # compute phase from group delay ------------------------------------------
+    sweep_ang = -1 * np.cumsum(sweep_gd) * 2 * np.pi * df
+
+    # wrap and correct phase to be real 0 at Nyquist
+    sweep_ang = pyfar.dsp.wrap_to_2pi(sweep_ang)
+    sweep_ang[sweep_ang > np.pi] -= 2*np.pi
+
+    if sweep_type == 'perfect_linear':
+        sweep_ang[-1] = 0
+    elif sweep_ang[-1] != 0 and not n_samples % 2:
+        factor = np.cumsum(np.ones_like(sweep_ang)) - 1
+        offset = df * sweep_ang[-1] / (sampling_rate / 2)
+        sweep_ang -= factor * offset
+        sweep_ang[-1] = np.abs(sweep_ang[-1])
+
+    # compute and finalize return data ----------------------------------------
+    # put group delay on pyfar FrequencyData
+    sweep_gd = pyfar.FrequencyData(
+        sweep_gd, pyfar.dsp.fft.rfftfreq(n_samples, sampling_rate))
+
+    # combine magnitude and phase of sweep
+    sweep = sweep_abs * np.exp(1j * sweep_ang)
+
+    # put sweep in pyfar.Signal an transform to time domain
+    sweep = pyfar.Signal(sweep, sampling_rate, n_samples, 'freq', 'none')
+
+    # find windowing end points heuristically. The start and stop margin are
+    # not reliable, because freq. domain synthesis causes pre/post-ringing
+    if fade_in or fade_out:
+        threshold = 10**(-60/20) * np.max(np.abs(sweep.time))
+        fade_start = np.argmax(np.abs(sweep.time) > threshold)
+        fade_end = n_samples - \
+            np.argmax(np.abs(sweep.time.flatten()[::-1]) > threshold)
+        print([fade_start, fade_end])
+
+    # generate windows for fade in and/or out
+    if fade_in and fade_out:
+        fade = [fade_start, fade_start + fade_in,
+                fade_end - fade_out, fade_end]
+        shape = 'symmetric'
+    elif fade_in:
+        fade = [fade_start, fade_start + fade_in]
+        shape = 'left'
+    elif fade_out:
+        fade = [fade_end - fade_out, fade_end]
+        shape = 'right'
+    else:
+        fade = None
+
+    # apply window
+    if fade is not None:
+        sweep = pyfar.dsp.time_window(sweep, fade, shape=shape)
+
+    # normalize to time domain amplitude of almost one
+    # (to avoid clipping if written to fixed point wav file)
+    sweep = pyfar.dsp.normalize(sweep) * (1 - 2**-15)
```

Comment:
```suggestion
    sweep = pyfar.dsp.normalize(sweep)
```
That is not something this function should take care of. That's the problem of a wav writer!

Reply:
agreed